### ISSUE TITLE
Docker container startup/restart resilience

### ISSUE THREAD

**Issue by beorn** [graphile/postgraphile#1041]
I'm submitting a ...

* [ ] bug report
* [x] feature request
* [ ] question 

PostGraphile version: v4.3.2

The current docker image fails if the database it is depending on isn't available, and when the database becomes unavailable.  This happens about 80% of the time when I use `docker-compose up` to startup the database and postgraphile together.  I get:

```log
Error: connect ECONNREFUSED 172.19.0.2:5432
TCPConnectWrap.afterConnect [as oncomplete] (net.js:1083:14)
```

In an orchestrated container type setup, you cannot really control the startup sequence very well, and the best practice is for containers to be resilient with respect to their backing services: to retry until they become available and to retry if they suddenly go missing (see [startup-order]).

It's possible to use the 3rd party tools as shown in [startup-order], but I think a better and more robust solution would be to have build this into postgraphile itself, to have it retry forever or N times, by default (my preferrence) or when a flag is set.

While the DB connection is not available, PostGraphile should probably not expose its external port either, since the service is unavailable.

[startup-order]: https://docs.docker.com/compose/startup-order/

**Comment by Vija02** [comment]
With docker-compose, you can use `depends_on` to make sure postgraphile runs after your database starts. It's also mentioned in the link you wrote. 
It's true that this might still fail. However, in my experience it's not a problem. 

Even when it is, you should use `restart: always` or the other variant to handle this.
Here's the reference for docker-compose:
https://docs.docker.com/compose/compose-file

**Comment by beorn** [comment]
Yes, I tried depends_on, and it doesn’t work; it’ll wait until the container has been started up, not until it’s ready, and postgres takes a while to become available, especially the first time, when postgres needs to initialize the database etc.  Also, depends_on only works with docker-compose, not in swarm (and presumably kubernetes).

The recommended way is to put reconnect logic into the apps/containers:
https://docs.docker.com/compose/startup-order/ <https://docs.docker.com/compose/startup-order/>

But `restart` works, as you mentioned, and I’m using that now, with `restart: unless-stopped`.

**Comment by benjie** [comment]
https://github.com/graphile/postgraphile/pull/1044

**Comment by benjie** [comment]
The new "retry on fail" behaviour introduced in #1044 (v4.4.0-beta.10) is being disabled by default in #1062 (which will be released with v4.4.0-rc.1); instead you must opt into the feature:

**To _not_ exit on init fail you need to set either `--retry-on-init-fail` on the CLI or `retryOnInitFail: true` in the library.**

Sorry for this inconvenience, but I wanted to ensure that 4.4 is not a breaking change for existing users who may be relying on the exit-on-fail behaviour of v4.0-v4.3.

rc1 should (hopefully) be out in the next few hours.
